File: ext/carray_attribute.c
/* ---------------------------------------------------------------------------

  Attribute and predicate accessors for CArray (obj_type / data_type
  / ndim / shape / dim / bytes / elements + entity? / virtual? /
  fixlen? / boolean? / read_only? / mask_array? / etc.) plus the
  view-chain navigators parent / root_array / ancestors.

  The corresponding YARD documentation lives in
  yard-stubs/carray_attribute.rb; this file carries only the
  implementation.

---------------------------------------------------------------------------- */

#include "carray.h"
#include "ca_obj_face.h"   /* CA_OBJ_RECORD + rb_ca_record_get_data_class */

/* ------------------------------------------------------------------- */

VALUE
rb_ca_obj_type (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return INT2NUM(ca->obj_type);
}

VALUE
rb_ca_data_type (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  /* Return a Symbol via the pre-computed ID cache; the C internal
     int8_t representation is unchanged. */
  return ID2SYM(ca_data_type_sym[ca->data_type]);
}

VALUE
rb_ca_ndim (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return INT2NUM(ca->ndim);
}

VALUE
rb_ca_bytes (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return SIZE2NUM(ca->bytes);
}

VALUE
rb_ca_flags (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return INT2NUM(ca->flags);
}

VALUE
rb_ca_elements (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return SIZE2NUM(ca->elements);
}

VALUE
rb_ca_dim (VALUE self)
{
  volatile VALUE dim;
  CArray *ca;
  int i;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  dim = rb_ary_new2(ca->ndim);
  for (i=0; i<ca->ndim; i++) {
    rb_ary_store(dim, i, SIZE2NUM(ca->dim[i]));
  }
  return dim;
}

VALUE
rb_ca_dim0 (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return SIZE2NUM(ca->dim[0]);
}

VALUE
rb_ca_dim1 (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ( ca->ndim >= 2 ) ? SIZE2NUM(ca->dim[1]) : Qnil;
}

VALUE
rb_ca_dim2 (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ( ca->ndim >= 3 ) ? SIZE2NUM(ca->dim[2]) : Qnil;
}

VALUE
rb_ca_dim3 (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ( ca->ndim >= 4 ) ? SIZE2NUM(ca->dim[3]) : Qnil;
}

VALUE
rb_ca_data_type_name (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return rb_str_new2(ca_type_name[ca->data_type]);
}

/* ------------------------------------------------------------------- */

int
ca_is_scalar (void *ap)
{
  CArray *ca = (CArray *) ap;
  return ca_test_flag(ca, CA_FLAG_SCALAR);
}

VALUE
rb_ca_is_scalar (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ( ca_is_scalar(ca) ) ? Qtrue : Qfalse;
}

VALUE
rb_obj_is_cscalar (VALUE obj)
{
  CArray *ca;
  if ( rb_obj_is_carray(obj) ) {
    TypedData_Get_Struct(obj, CArray, &carray_data_type, ca);
    return ( ca_is_scalar(ca) ) ? Qtrue : Qfalse;
  }
  return Qfalse;
}

/* ------------------------------------------------------------------- */

int
ca_is_view (void *ap)
{
  CArray *ca = (CArray *) ap;
  return ( ca_func[ca->obj_type].entity_type == CA_VIEW_ARRAY ) ? 1 : 0;
}

VALUE
rb_ca_is_entity (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ( ca_is_view(ca) ) ? Qfalse : Qtrue;
}

VALUE
rb_ca_is_virtual (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ( ca_is_view(ca) ) ? Qtrue : Qfalse;
}

/* ------------------------------------------------------------------- */

VALUE
rb_ca_is_attached (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ( ca_is_attached(ca) ) ? Qtrue : Qfalse;
}

/* ------------------------------------------------------------------- */

VALUE
rb_ca_is_empty (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ( ca->elements == 0 ) ? Qtrue : Qfalse;
}

/* ------------------------------------------------------------------- */

int
ca_is_readonly (void *ap)
{
  CArray *ca = (CArray *) ap;
  if ( ca_test_flag(ca, CA_FLAG_READ_ONLY) ) {           /* test -> true */
    return 1;
  }
  else if ( ca_test_flag(ca, CA_FLAG_MULTI_PARENTS) ) {
    /* multi-parent view: read-only iff ANY parent is (a whole-view write
       cannot be honoured if any backing parent rejects writes). */
    CAMultiParent *mp = (CAMultiParent *) ca;
    int32_t k;
    for ( k = 0; k < mp->n_parents; k++ ) {
      if ( ca_is_readonly(mp->parents[k]) ) {
        ca_set_flag(ca, CA_FLAG_READ_ONLY);
        return 1;
      }
    }
    return 0;
  }
  else {                                                 /* test -> false */
    if ( ca_is_view(ca) && CAVIEW(ca)->parent ) {
      if ( ca_is_readonly(CAVIEW(ca)->parent) ) {     /* test -> true */
        ca_set_flag(ca, CA_FLAG_READ_ONLY);
        return 1;
      }
    }
    return 0;                                            /* all test -> false */
  }
}

VALUE
rb_ca_is_read_only (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ( ca_is_readonly(ca) ) ? Qtrue : Qfalse;
}

/* ------------------------------------------------------------------- */

int
ca_is_mask_array (void *ap)
{
  CArray *ca = (CArray *) ap;
  if ( ca_test_flag(ca, CA_FLAG_MASK_ARRAY) ) {           /* test -> true */
    return 1;
  }
  else {                                                   /* test -> false */
    /* Identity property: a CAStack is not a mask array unless
       flagged directly (the internal CAStackMask is).  Do not
       inherit the property across a multi-parent fan-out — stop at
       the boundary. */
    if ( ca_is_view(ca) && CAVIEW(ca)->parent
         && ! ca_test_flag(ca, CA_FLAG_MULTI_PARENTS) ) {
      if ( ca_is_mask_array(CAVIEW(ca)->parent) ) {     /* test -> true */
        ca_set_flag(ca, CA_FLAG_MASK_ARRAY);
        return 1;
      }
    }
    return 0;                                            /* all test -> false */
  }
}

VALUE
rb_ca_is_mask_array (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ( ca_is_mask_array(ca) ) ? Qtrue : Qfalse;
}

/* ------------------------------------------------------------------- */

int
ca_is_value_array (void *ap)
{
  CArray *ca = (CArray *) ap;
  if ( ca_test_flag(ca, CA_FLAG_VALUE_ARRAY) ) {           /* test -> true */
    return 1;
  }
  else {                                                   /* test -> false */
    /* Identity property: a CAStack is never "the .value of" a
       single array (s.value wraps it in a flagged CARefer).  Do
       not inherit across a multi-parent fan-out — stop at the
       boundary. */
    if ( ca_is_view(ca) && CAVIEW(ca)->parent
         && ! ca_test_flag(ca, CA_FLAG_MULTI_PARENTS) ) {
      if ( ca_is_value_array(CAVIEW(ca)->parent) ) {     /* test -> true */
        ca_set_flag(ca, CA_FLAG_VALUE_ARRAY);
        return 1;
      }
    }
    return 0;                                            /* all test -> false */
  }
}

VALUE
rb_ca_is_value_array (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ( ca_is_value_array(ca) ) ? Qtrue : Qfalse;
}

/* ------------------------------------------------------------------- */

VALUE
rb_ca_is_face (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ( ca_is_face(ca) ) ? Qtrue : Qfalse;
}

/* ------------------------------------------------------------------- */

int
ca_is_fixlen_type (void *ap)
{
  CArray *ca = (CArray *) ap;
  return ( ca->data_type == CA_FIXLEN );
}

VALUE
rb_ca_is_fixlen_type (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ca_is_fixlen_type(ca) ? Qtrue : Qfalse;
}

/* ------------------------------------------------------------------- */

int
ca_is_boolean_type (void *ap)
{
  CArray *ca = (CArray *) ap;
  return ( ca->data_type == CA_BOOLEAN );
}

VALUE
rb_ca_is_boolean_type (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ca_is_boolean_type(ca) ? Qtrue : Qfalse;
}

/* ------------------------------------------------------------------- */

int
ca_is_numeric_type (void *ap)
{
  CArray *ca = (CArray *) ap;
  return ( ( ca->data_type >= CA_INT8 ) &&
           ( ca->data_type <= CA_CMPLX256 ) );
}

VALUE
rb_ca_is_numeric_type (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ca_is_numeric_type(ca) ? Qtrue : Qfalse;
}

/* ------------------------------------------------------------------- */

int
ca_is_integer_type (void *ap)
{
  CArray *ca = (CArray *) ap;
  return ( ( ca->data_type >= CA_INT8 ) &&
           ( ca->data_type <= CA_UINT64 ) );
}

VALUE
rb_ca_is_integer_type (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ca_is_integer_type(ca) ? Qtrue : Qfalse;
}

/* ------------------------------------------------------------------- */

int
ca_is_float_type (void *ap)
{
  CArray *ca = (CArray *) ap;
  return ( ( ca->data_type >= CA_FLOAT32 ) &&
           ( ca->data_type <= CA_FLOAT128 ) );
}

VALUE
rb_ca_is_float_type (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ca_is_float_type(ca) ? Qtrue : Qfalse;
}

/* ------------------------------------------------------------------- */

int
ca_is_complex_type (void *ap)
{
  CArray *ca = (CArray *) ap;
  return ( ( ca->data_type >= CA_CMPLX64 ) &&
           ( ca->data_type <= CA_CMPLX256 ) );
}

VALUE
rb_ca_is_complex_type (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ca_is_complex_type(ca) ? Qtrue : Qfalse;
}

/* ------------------------------------------------------------------- */

int
ca_is_object_type (void *ap)
{
  CArray *ca = (CArray *) ap;
  return ( ca->data_type == CA_OBJECT );
}

VALUE
rb_ca_is_object_type (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  return ca_is_object_type(ca) ? Qtrue : Qfalse;
}

/* ------------------------------------------------------------------- */

static ID id_parent;

VALUE
rb_ca_parent (VALUE self)
{
  return rb_ivar_get(self, id_parent);
}

VALUE
rb_ca_set_parent (VALUE self, VALUE obj)
{
  rb_ivar_set(self, id_parent, obj);
  if ( OBJ_FROZEN(obj) ) {
    rb_ca_freeze(self);
  }
  return obj;
}

/* ------------------------------------------------------------------- */

VALUE
rb_ca_data_class (VALUE self)
{
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  /* Only a Face carries data_class; a non-Face CArray is always
     Qnil.  CARecord stores it in its tail (reached directly via
     obj_type, no Ruby roundtrip); numeric Faces such as
     CATime / CATimedelta have no data_class. */
  if ( ! ca_test_flag(ca, CA_FLAG_IS_FACE) ) {
    return Qnil;
  }
  if ( ca->obj_type == CA_OBJ_RECORD ) {
    return rb_ca_record_get_data_class(ca);
  }
  return Qnil;
}

VALUE
rb_ca_has_data_class (VALUE self)
{
  return RTEST(rb_ca_data_class(self)) ? Qtrue : Qfalse;
}

/* ------------------------------------------------------------------- */

/* data_class= is defined but always raises.  data_class now lives
   on the Face (CARecord) tail; there is no way to attach one to a
   non-Face CArray in-place.  The setter is kept only so a call
   surfaces the migration message rather than NoMethodError.
   NORETURN: rb_raise never returns, so neither does this function. */
NORETURN(VALUE rb_ca_set_data_class(VALUE self, VALUE klass));
VALUE
rb_ca_set_data_class (VALUE self, VALUE klass)
{
  rb_raise(rb_eArgError,
    "CArray#data_class= was removed in 3.0. "
    "Use `CARecord.new(klass, *shape)` or `CARecord.wrap(entity, klass)`.");
}

/* ------------------------------------------------------------------- */

CArray *
ca_root_array (void *ap)
{
  CArray *ca = (CArray *)ap;
  if ( ca_is_entity(ca) ) {
    return ca;
  }
  else if ( ca_test_flag(ca, CA_FLAG_MULTI_PARENTS) ) {
    /* Multi-parent view fans out to K roots, so the single-
       reference chain terminates here — the view itself is the
       root boundary. */
    return ca;
  }
  else {
    CAView *cr = (CAView *)ca;
    if ( ! cr->parent ) {
      return ca;
    }
    else {
      return ca_root_array(cr->parent);
    }
  }
}

static VALUE
rb_ca_root_array (VALUE self)
{
  volatile VALUE refary;
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  if ( ca_is_entity(ca) ) {
    return self;
  }
  else if ( ca_test_flag(ca, CA_FLAG_MULTI_PARENTS) ) {
    return self;   /* Multi-parent: root boundary (fans out to K roots). */
  }
  else {
    refary = rb_ca_parent(self);
    if ( NIL_P(refary) ) {
      return self;
    }
    else {
      return rb_ca_root_array(refary);
    }
  }
}

/* ------------------------------------------------------------------- */

static VALUE
rb_ca_ancestors_loop (VALUE self, VALUE list)
{
  volatile VALUE refary;
  CArray *ca;
  TypedData_Get_Struct(self, CArray, &carray_data_type, ca);
  rb_ary_unshift(list, self);
  if ( ca_is_entity(ca) ) {
    return list;
  }
  else if ( ca_test_flag(ca, CA_FLAG_MULTI_PARENTS) ) {
    return list;   /* Multi-parent: chain terminates here (fans out to K). */
  }
  else {
    refary = rb_ca_parent(self);
    if ( rb_obj_is_carray(refary) ) {
      return rb_ca_ancestors_loop(refary, list);
    }
    else {
      return list;
    }
  }
}

static VALUE
rb_ca_ancestors (VALUE self)
{
  volatile VALUE list;
  list = rb_ary_new();
  return rb_ca_ancestors_loop(self, list);
}

/* ------------------------------------------------------------------- */

void
Init_carray_attribute (void)
{
  id_parent     = rb_intern("parent");

  rb_define_method(rb_cCArray, "obj_type", rb_ca_obj_type, 0);
  rb_define_method(rb_cCArray, "data_type", rb_ca_data_type, 0);
  rb_define_method(rb_cCArray, "bytes", rb_ca_bytes, 0);
  rb_define_method(rb_cCArray, "flags", rb_ca_flags, 0);
  rb_define_method(rb_cCArray, "ndim", rb_ca_ndim, 0);
  rb_define_method(rb_cCArray, "rank", rb_ca_ndim, 0);
  rb_define_method(rb_cCArray, "shape", rb_ca_dim, 0);
  rb_define_method(rb_cCArray, "dim", rb_ca_dim, 0);
  rb_define_method(rb_cCArray, "dim0", rb_ca_dim0, 0);
  rb_define_method(rb_cCArray, "dim1", rb_ca_dim1, 0);
  rb_define_method(rb_cCArray, "dim2", rb_ca_dim2, 0);
  rb_define_method(rb_cCArray, "dim3", rb_ca_dim3, 0);
  rb_define_method(rb_cCArray, "elements", rb_ca_elements, 0);
  rb_define_method(rb_cCArray, "length", rb_ca_elements, 0); 
  rb_define_method(rb_cCArray, "size", rb_ca_elements, 0); 
  
  rb_define_method(rb_cCArray, "data_type_name", rb_ca_data_type_name, 0);

  rb_define_method(rb_cCArray, "parent", rb_ca_parent, 0);

  rb_define_method(rb_cCArray, "data_class", rb_ca_data_class, 0);
  rb_define_method(rb_cCArray, "data_class=", rb_ca_set_data_class, 1);

  rb_define_method(rb_cCArray, "scalar?", rb_ca_is_scalar, 0);

  rb_define_method(rb_cCArray, "entity?", rb_ca_is_entity, 0);
  rb_define_method(rb_cCArray, "virtual?", rb_ca_is_virtual, 0);
  rb_define_method(rb_cCArray, "value_array?", rb_ca_is_value_array, 0);
  rb_define_method(rb_cCArray, "mask_array?", rb_ca_is_mask_array, 0);
  rb_define_method(rb_cCArray, "face?", rb_ca_is_face, 0);

  rb_define_method(rb_cCArray, "empty?", rb_ca_is_empty, 0);
  rb_define_method(rb_cCArray, "read_only?", rb_ca_is_read_only, 0);
  rb_define_method(rb_cCArray, "attached?", rb_ca_is_attached, 0);

  rb_define_method(rb_cCArray, "has_data_class?", rb_ca_has_data_class, 0);

  rb_define_method(rb_cCArray, "fixlen?",   rb_ca_is_fixlen_type, 0);
  rb_define_method(rb_cCArray, "boolean?",  rb_ca_is_boolean_type, 0);
  rb_define_method(rb_cCArray, "numeric?",  rb_ca_is_numeric_type, 0);
  rb_define_method(rb_cCArray, "integer?",  rb_ca_is_integer_type, 0);
  rb_define_method(rb_cCArray, "float?",    rb_ca_is_float_type, 0);
  rb_define_method(rb_cCArray, "complex?",  rb_ca_is_complex_type, 0);
  rb_define_method(rb_cCArray, "object?",   rb_ca_is_object_type, 0);

  rb_define_method(rb_cCArray, "root_array", rb_ca_root_array, 0);
  rb_define_method(rb_cCArray, "ancestors", rb_ca_ancestors, 0);
}

File: yard-stubs/carray_attribute.rb
# DOCUMENTATION ONLY — DO NOT REQUIRE.
# Stubs for methods defined in ext/carray_attribute.c.
# See yard-stubs/README.md and yard-stubs/STYLE.md.

class CArray
  # @!group Attributes

  # @overload obj_type
  #   Returns the object-type integer of `self` (e.g. `CA_OBJ_ARRAY`,
  #   `CA_OBJ_BLOCK`).
  #
  #   Rarely needed in user code; the same information is available
  #   from `self.class`.
  #   @return [Integer]
  def obj_type; end

  # @!group Type inquiry

  # @overload data_type
  #   Returns the data type of each element as a Symbol (e.g. `:int32`,
  #   `:float64`).
  #   @return [Symbol]
  def data_type; end

  # @overload data_type_name
  #   Returns the String name of `data_type` (e.g. `"int32"`,
  #   `"fixlen"`).
  #   @return [String]
  def data_type_name; end

  # @!group Attributes

  # @overload ndim
  #   Returns the number of dimensions of `self`.
  #   @return [Integer]
  def ndim; end

  # @overload rank
  #   @deprecated Use {#ndim}.
  #   @return [Integer]
  def rank; end

  # @overload bytes
  #   Returns the byte size of one element. Equals
  #   `CArray.sizeof(data_type)` for numeric types; for fixed-length
  #   types this method is the only way to obtain the size.
  #   @return [Integer]
  def bytes; end

  # @overload flags
  #   Returns the raw internal flag bitset (`ca->flags`) as an Integer.
  #   The bits record structural facts about the array (scalar, Face,
  #   view, read-only, …). This is a low-level provenance value with no
  #   stable public meaning; it is exposed mainly so the `_CARRAY3`
  #   serializer can snapshot it into a saved file's header.
  #   @return [Integer]
  def flags; end

  # @overload elements
  #   Returns the total number of elements (product of `shape`).
  #   @return [Integer]
  def elements; end

  # @overload size
  #   Alias of {#elements}.
  #   @return [Integer]
  def size; end

  # @overload length
  #   @deprecated Use {#elements} or {#size}.
  #   @return [Integer]
  def length; end

  # @overload shape
  #   Returns a freshly allocated Array containing the dimensional
  #   shape of `self` (e.g. `[2, 3]` for a 2×3 array).
  #
  #   This is the recommended accessor. {#dim} is a legacy alias.
  #   @return [Array<Integer>]
  #   @example
  #     CArray.float64(2, 3).shape  # => [2, 3]
  def shape; end

  # @overload dim
  #   @deprecated Use {#shape}.
  #   @return [Array<Integer>]
  def dim; end

  # @overload dim0
  #   Returns `shape[0]`.
  #   @return [Integer]
  def dim0; end

  # @overload dim1
  #   Returns `shape[1]`, or `nil` if `ndim < 2`.
  #   @return [Integer, nil]
  def dim1; end

  # @overload dim2
  #   Returns `shape[2]`, or `nil` if `ndim < 3`.
  #   @return [Integer, nil]
  def dim2; end

  # @overload dim3
  #   Returns `shape[3]`, or `nil` if `ndim < 4`.
  #   @return [Integer, nil]
  def dim3; end

  # @overload parent
  #   Returns the parent CArray of `self`, or `nil` if `self` has no
  #   parent (i.e. is an entity).
  #   @return [CArray, nil]
  def parent; end

  # @overload root_array
  #   Returns the array at the root of the view chain (the entity).
  #   Returns `self` if `self` is already an entity.
  #   @return [CArray]
  def root_array; end

  # @overload ancestors
  #   Returns the list of arrays in the view chain, ordered from root
  #   to `self`.
  #   @return [Array<CArray>]
  def ancestors; end

  # @!group Type inquiry

  # @overload data_class
  #   Returns the `data_class` of `self` if it is a Face that carries
  #   one (e.g. `CARecord`). Returns `nil` for non-Face arrays.
  #   @return [Class, nil]
  def data_class; end

  # @overload data_class=(klass)
  #   Always raises. `data_class` now lives on the Face tail
  #   (`CARecord`); use `CARecord.new(klass, *shape)` or
  #   `CARecord.wrap(entity, klass)` instead.
  #   @param klass [Class] unused.
  #   @return [void]
  #   @raise [ArgumentError] always.
  def data_class=(klass); end

  # @!endgroup

  # @!group State inquiry

  # @overload scalar?
  #   Returns `true` if `self` is a {CScalar}.
  #   @return [Boolean]
  def scalar?; end

  # @overload entity?
  #   Returns `true` if `self` is an entity array (not a virtual
  #   view).
  #   @return [Boolean]
  def entity?; end

  # @overload virtual?
  #   Returns `true` if `self` is a virtual array (a view, not an
  #   entity).
  #   @return [Boolean]
  def virtual?; end

  # @overload attached?
  #   Returns `true` if `self` is currently attached
  #   (`ptr != NULL`).
  #   @return [Boolean]
  def attached?; end

  # @overload empty?
  #   Returns `true` if `self` has zero elements.
  #   @return [Boolean]
  def empty?; end

  # @overload read_only?
  #   Returns `true` if `self` is read-only.
  #   @return [Boolean]
  def read_only?; end

  # @overload mask_array?
  #   Returns `true` if `self` is itself a mask array. (Not the same
  #   as "a masked array"; for that see {#has_mask?}.)
  #   @return [Boolean]
  def mask_array?; end

  # @overload value_array?
  #   Returns `true` if `self` is a value array (the `.value` view of
  #   a masked array).
  #   @return [Boolean]
  def value_array?; end

  # @!group Type inquiry

  # @overload face?
  #   Returns `true` if `self` is a Face view (`CA_FLAG_IS_FACE`
  #   set).
  #   @return [Boolean]
  def face?; end

  # @overload has_data_class?
  #   Returns `true` if `self` carries a `data_class` (i.e. is a Face
  #   such as `CARecord`).
  #   @return [Boolean]
  def has_data_class?; end

  # @overload fixlen?
  #   Returns `true` if `self` is a fixed-length type array.
  #   @return [Boolean]
  def fixlen?; end

  # @overload boolean?
  #   Returns `true` if `self` is a boolean type array.
  #   @return [Boolean]
  def boolean?; end

  # @overload numeric?
  #   Returns `true` if `self` is a numeric type array (any integer,
  #   float, or complex type).
  #   @return [Boolean]
  def numeric?; end

  # @overload integer?
  #   Returns `true` if `self` is an integer type array (signed or
  #   unsigned, any width).
  #   @return [Boolean]
  def integer?; end

  # @overload float?
  #   Returns `true` if `self` is a floating-point type array.
  #   @return [Boolean]
  def float?; end

  # @overload complex?
  #   Returns `true` if `self` is a complex type array.
  #   @return [Boolean]
  def complex?; end

  # @overload object?
  #   Returns `true` if `self` is an object type array
  #   (`data_type == :object`).
  #   @return [Boolean]
  def object?; end

  # @!endgroup
end
